interfaces/devdoc/file_requirements.md

```diff
@@ -0,0 +1,152 @@
+# file
+================
+
+## Overview
+
+The `file` module provides a platform-independent API for asynchronous file operations.
+
+-`file_create`: returns a file handle for the given file name.
+-`file_destroy`: closes the given file handle.
+-`file_write_async`: enqueues an asynchronous write request for a file at a given position.
+-`file_read_async`: enqueues an asynchronous read request for a file at a given position and size.
+-`file_extend`: expands the given file to be of desired size.
+
+## Exposed API
+
+```c
+#define FILE_WRITE_ASYNC_VALUES \
+    FILE_WRITE_ASYNC_INVALID_ARGS, \
+    FILE_WRITE_ASYNC_WRITE_ERROR, \
+    FILE_WRITE_ASYNC_ERROR,\
+    FILE_WRITE_ASYNC_OK
+MU_DEFINE_ENUM(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_VALUES);
+
+#define FILE_READ_ASYNC_VALUES \
+    FILE_READ_ASYNC_INVALID_ARGS, \
+    FILE_READ_ASYNC_READ_ERROR, \
+    FILE_READ_ASYNC_ERROR,\
+    FILE_READ_ASYNC_OK
+MU_DEFINE_ENUM(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_VALUES);
+
+typedef struct FILE_HANDLE_DATA_TAG* FILE_HANDLE;
+typedef void(*FILE_REPORT_FAULT)(void* user_report_fault_context, const char* information);
+
+typedef void(*FILE_WRITE_CB)(void* user_context, bool is_successful);
+typedef void(*FILE_READ_CB)(void* user_context, bool is_successful, CONSTBUFFER_HANDLE content);
+
+MOCKABLE_FUNCTION(, FILE_HANDLE, file_create, EXECUTION_ENGINE_HANDLE, execution_engine, const char*, full_file_name, FILE_REPORT_FAULT, user_report_fault_callback, void*, user_report_fault_context);
+MOCKABLE_FUNCTION(, void, file_destroy, FILE_HANDLE, handle);
+
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_WRITE_ASYNC_RESULT, file_write_async, FILE_HANDLE, handle, CONSTBUFFER_HANDLE, source, uint64_t, position, FILE_WRITE_CB, user_callback, void*, user_context)(FILE_WRITE_ASYNC_OK, FILE_WRITE_ASYNC_ERROR);

```

Comment:
>CONSTBUFFER_HANDLE [](start = 97, length = 18)

I still believe these functions should not take constbuffer_handles, rather they can take const unsigned char*, size_t size. Reason is bleeding of higher abstraction concepts in something called "pal".

Reply:
Changed to char* buffers in next iteration

---
In reply to: [452448920](https://github.com/Azure/azure-c-pal/pull/7#discussion_r452448920) [](ancestors = 452448920)